src/bytes_ser_de.rs

```diff
@@ -68,22 +74,18 @@ impl<'a> Deserializer<'a> {
     /// Reads a `u64` from the `Deserializer`.
     #[inline]
     pub fn read_u64(&mut self) -> Result<u64, CryptoCoreError> {
-        leb128::read::unsigned(&mut self.readable).map_err(|e| {
-            CryptoCoreError::InvalidSize(format!(
-                "Deserializer: failed reading the size of the next array: {}",
-                e
-            ))
-        })
+        leb128::read::unsigned(&mut self.readable).map_err(CryptoCoreError::ReadLeb128Error)
     }
 
     /// Reads an array of bytes of length `LENGTH` from the `Deserializer`.
     #[inline]
     pub fn read_array<const LENGTH: usize>(&mut self) -> Result<[u8; LENGTH], CryptoCoreError> {
         let mut buf = [0; LENGTH];
         self.readable.read_exact(&mut buf).map_err(|_| {
-            CryptoCoreError::InvalidSize(format!(
-                "Deserializer: failed reading array of: {LENGTH} bytes",
-            ))
+            CryptoCoreError::DeserialisationSizeError {
+                given: self.readable.len(),
+                expected: LENGTH,
+            }
```

Comment:
Are you sure an error on the size is the only possible kind of error? It is not clear to me when reading the doc of `read_exact()`:
```
If this function encounters an error of the kind
[`ErrorKind::Interrupted`](https://doc.rust-lang.org/nightly/std/io/error/enum.ErrorKind.html) then the error is ignored and the operation
will continue.

If this function encounters an "end of file" before completely filling
the buffer, it returns an error of the kind [`ErrorKind::UnexpectedEof`](https://doc.rust-lang.org/nightly/std/io/error/enum.ErrorKind.html).
The contents of `buf` are unspecified in this case.

If any other read error is encountered then this function immediately
returns. The contents of `buf` are unspecified in this case.

If this function returns an error, it is unspecified how many bytes it
has read, but it will never read more than would be necessary to
completely fill the buffer.
```

Reply:
`read_exact` has only one `Err` case from the code I think…

```
 #[inline]
    fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
        if buf.len() > self.len() {
            return Err(io::const_io_error!(
                ErrorKind::UnexpectedEof,
                "failed to fill whole buffer"
            ));
        }
        let (a, b) = self.split_at(buf.len());

        // First check if the amount of bytes we want to read is small:
        // `copy_from_slice` will generally expand to a call to `memcpy`, and
        // for a single byte the overhead is significant.
        if buf.len() == 1 {
            buf[0] = a[0];
        } else {
            buf.copy_from_slice(a);
        }

        *self = b;
        Ok(())
    }
```